### DISCUSSION TITLE
Voltage Source

### DISCUSSION THREAD

**Discussion by wktanghk** [scikit-rf/scikit-rf#759] [Q&A] [answered]
Hi, I would like to know if it is possible to apply a voltage swing on one port, and measure the power on another port. Looking through the Circuit documentation, there is a nice function 'voltages_external'. The input stimulus is a power souce at one port, and the observed outputs are the voltages & currents on other ports.

In my case, I would like to do the opposite which is similar to an AC simulation. The input stimulus is a voltage source with a given input amplitude (e.g. 1V, 0phase) across frequency. And measure the delivered power on another port [i.e. 0.5x Re (VI*) flowing out at the output port].

One example could be a 2-port transformer. In a traditional schematic test bench, I can apply a Vac (AC voltage source) across the primary differential port. This would be the only connection on the input side (i.e. no other termination). On the output side, it would be loaded with some arbitrary load (e.g. 50ohm). And I could then measure the power delivering to this load by monitoring the voltage across the load & the current flowing into the load.

I wonder if there are existing functions in scikit-rf that can handle this. Or if there are other functions I can get started with, I would really appreciate if someone could recommend. Thanks!
#257 #381

**Comment by jhillairet** [comment]
Presently no function is intended to do that directly, but it seems doable.

A way to do that could be the following (to be tested):
- calculate the excitation power wave associated to your voltage (not done)
- calculate the internal s parameters of the circuit (ok)
- calculate voltages and currents at the external ports (ok) and deduce the
power (not done)

I do not have access to a computer right now, but if you send a numerical
example with a lumped circuit, I can test in the future.

**Comment by wktanghk** [comment, marked as answer]
Hi @jhillairet, thank you very much for confirming and your recommendation. For your suggested steps, step 2 & 3 are relatively straight forward. For step 1 where we calculate the excitation power wave for a given voltage stimulus, I'm not sure if the fact that we don't have a specified source impedance would complicate the power calculation. Since I'm applying an AC voltage source at the input port as a stimulus, instead of connecting an actual port with a certain impedance. I'm not sure if I should still connect a Circuit Port in Scikit-rf and set its port impedance to 0 (?)

I will also experiment on my end.

For your reference, a lump circuit I was testing is a lumped transformer. Primary is 1nH + 1ohm; Secondary is 2nH + 2ohm; K = 0.75. This is a 2-port network (Port1 is differential across the primary plus & minus; Port2 is differential across the secondary plus & minus). I then apply a 1V peak voltage at Port1, and a complex load (20-j20) at Port2. From this AC simulation @ 1GHz, I measure the voltage & current at Port2 as follows:

V = (1.09 -j0.08)V
I = (0.029 + j0.025)A [current flowing into the load]

Finally, to find the Power delivered to the load at Port2, I used P = 0.5 x VI*, which gives 0.015W

I was looking for ways to attach the S2P & snapshots of the schematic & waveforms, but couldn't find a way to do it here. 

Thanks again. And I will look into building the code based on your recommendation as well.

Thanks,
William

**Reply by wktanghk** [reply]
Hi @jhillairet, just to close the loop on this. Upon further experiments, I think it might be a little more straightforward to directly derive the V & I at the output port by converting to y-parameters, and excite with the voltage stimulus. This issue is now closed. Thanks for helping to take a look into this and your suggestion. 

Thanks!
William

**Reply by jhillairet** [reply]
Dear @wktanghk , 

If you have some, could you copy/paste an example?

**Reply by wktanghk** [reply]
Hi @jhillairet ,

Absolutely. Let me clean up paste a simplified example here.
